controller/apps.go

```diff
@@ -0,0 +1,464 @@
+	}
+}
+
+func tostring(item interface{}) string {
+	bytes, err := json.MarshalIndent(item, "", "  ")
+	if err != nil {
+		return err.Error()
+	}
+	return string(bytes)
+}
+
+func getAndCheckOSIOClient(ctx context.Context) *OSIOClient {
+
+	// defaults
+	host := "localhost"
+	scheme := "https"
+
+	req := goa.ContextRequest(ctx)
+	if req != nil {
+		// Note - it's probably more efficient to force a loopback host, and only use the port number here
+		// (on some systems using a non-loopback interface forces a network stack traverse)
+		host = req.Host
+		scheme = req.URL.Scheme
+	}
+
+	if os.Getenv("FABRIC8_WIT_API_URL") != "" {
+		witurl, err := url.Parse(os.Getenv("FABRIC8_WIT_API_URL"))
+		if err != nil {
+			log.Warn(ctx, nil, "Cannot parse FABRIC8_WIT_API_URL; assuming localhost")
+		}
+		host = witurl.Host
+		scheme = witurl.Scheme
+	}
+
+	oc := NewOSIOClient(ctx, scheme, host)
+
+	return oc
+}
+
+func (c *AppsController) getSpaceNameFromSpaceID(ctx context.Context, spaceID uuid.UUID) (*string, error) {
+	// TODO - add a cache in AppsController - but will break if user can change space name
+	// use WIT API to convert Space UUID to Space name
+	osioclient := getAndCheckOSIOClient(ctx)
+
+	osioSpace, err := osioclient.GetSpaceByID(ctx, spaceID)
+	if err != nil {
+		return nil, err
+	}
+	return osioSpace.Attributes.Name, nil
+}
+
+func getNamespaceName(ctx context.Context) (*string, error) {
+
+	osioclient := getAndCheckOSIOClient(ctx)
+	kubeSpaceAttr, err := osioclient.GetNamespaceByType(ctx, nil, "user")
+	if err != nil {
+		return nil, err
+	}
+	if kubeSpaceAttr == nil || kubeSpaceAttr.Name == nil {
+		return nil, witerrors.NewNotFoundError("namespace", "user")
+	}
+
+	return kubeSpaceAttr.Name, nil
+}
+
+func getUser(authClient authservice.Client, ctx context.Context) (*authservice.User, error) {
+	// get the user definition (for cluster URL)
+	resp, err := authClient.ShowUser(ctx, authservice.ShowUserPath(), nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, fmt.Errorf("Failed to GET user due to status code %d", status)
+	}
+
+	var respType authservice.User
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
+	}
+	return &respType, nil
+}
+
+func getTokenData(authClient authservice.Client, ctx context.Context, forService string) (*authservice.TokenData, error) {
+
+	resp, err := authClient.RetrieveToken(ctx, authservice.RetrieveTokenPath(), forService, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	status := resp.StatusCode
+	if status < 200 || status > 300 {
+		return nil, errors.New("Failed to GET user due to status code " + string(status))
+	}
+
+	var respType authservice.TokenData
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		return nil, err
```

Comment:
Please use `errs.WithStack(err)` when returning plain errors.

Reply:
Done (in followup PR).
A style question - should I do this even when I know the call that caused the error already has a stack (because it's from my code)?  I have assumed the answer is 'yes'.